rescan_test.go

```diff
@@ -0,0 +1,621 @@
+			// retrying a previous one.
+			name: "new block after retry",
+			run: func(ctx *rescanTestContext) {
+				// Override the GetCFilter implementation to
+				// mark the following blocks as retries.
+				ctx.chain.overrideGetCFilter(filterNotFound)
+
+				// Adding new blocks to the chain should result
+				// in the first retry block being queried until
+				// successful.
+				block1 := ctx.chain.addNewBlock(true)
+				assertFilterQueried(ctx.t, block1.Hash)
+				block2 := ctx.chain.addNewBlock(true)
+				assertFilterQueried(ctx.t, block1.Hash)
+
+				// Revert back to the original GetCFilter
+				// implementation, causing us to properly notify
+				// the blocks in order.
+				ctx.chain.overrideGetCFilter(nil)
+
+				ctx.recvBlockConnected(block1.Height)
+				ctx.recvBlockConnected(block2.Height)
+			},
+		},
+		{
+			// We should stop retrying blocks for which we've
+			// received disconnected notifications for.
+			name: "reorg n-1 retry blocks",
+			run: func(ctx *rescanTestContext) {
+				// Override the GetCFilter implementation to
+				// mark the following blocks as retries.
+				ctx.chain.overrideGetCFilter(filterNotFound)
+
+				// Adding new blocks to the chain should result
+				// in the first retry block being queried until
+				// successful.
+				block := ctx.chain.addNewBlock(true)
+				assertFilterQueried(ctx.t, block.Hash)
+				ctx.chain.addNewBlock(true)
+				assertFilterQueried(ctx.t, block.Hash)
+				ctx.chain.addNewBlock(true)
+				assertFilterQueried(ctx.t, block.Hash)
+
+				// We'll then reorg back to the first retry
+				// block. This should cause us to stop retrying
+				// the reorged blocks.
+				ctx.chain.rollbackToHeight(block.Height, true)
+
+				// We should then receive a notification for it
+				// once we revert back to the original
+				// GetCFilter implementation.
+				ctx.chain.overrideGetCFilter(nil)
+				ctx.recvBlockConnected(block.Height)
+			},
+		},
+		{
+			// We should stop retrying blocks for which we've
+			// received disconnected notifications for.
+			name: "reorg all retry blocks",
+			run: func(ctx *rescanTestContext) {
+				// Override the GetCFilter implementation to
+				// mark the following blocks as retries.
+				ctx.chain.overrideGetCFilter(filterNotFound)
+
+				// Note the block which we'll revert back to.
+				bestBlock, err := ctx.chain.BestBlock()
+				if err != nil {
+					ctx.t.Fatalf("unable to retrieve "+
+						"best block: %v", err)
+				}
+
+				// Adding new blocks to the chain should result
+				// in the first retry block being queried until
+				// successful.
+				block1 := ctx.chain.addNewBlock(true)
+				assertFilterQueried(ctx.t, block1.Hash)
+				ctx.chain.addNewBlock(true)
+				assertFilterQueried(ctx.t, block1.Hash)
+
+				// We'll then reorg back to a height where the
+				// new blocks we've added and retried are stale,
+				// causing us to no longer retry them.
+				ctx.chain.rollbackToHeight(bestBlock.Height, true)
```

Comment:
Shouldn't we verify that the next block added (for a new chain) causes us to continue fetching down that new chain?

Reply:
Good point, done.